language_service/src/identifier_locator.rs

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::mem::replace;
+use std::rc::Rc;
+
+use crate::qsc_utils::{find_ident, find_item, span_contains, span_touches, Compilation};
+use qsc::ast::visit::{walk_expr, walk_namespace, walk_pat, walk_ty_def, Visitor};
+use qsc::{ast, hir, resolve};
+
+#[allow(unused_variables)]
+pub(crate) trait LocatorAPI<'package> {
+    fn at_callable_def(
+        &mut self,
+        context: &LocatorContext<'package>,
+        decl: &'package ast::CallableDecl,
+    ) {
+    }
+
+    fn at_callable_ref(
+        &mut self,
+        path: &'package ast::Path,
+        item_id: &'package hir::ItemId,
+        item: &'package hir::Item,
```

Comment:
Hmm, it looks like the `item` is only used to get the containing namespace. Personally I think it would look cleaner to just resolve the namespace in the locator and pass it here. Even though only hover uses it.

Reply:
Discussed offline: Since Hover is the only one to use this, let's wait to see if any other features added in the future need this info. In the meantime it can live in the hover.